File: leetcode/cpp_solutions/merge_two_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {

        if (list1 == nullptr && list2 == nullptr) return nullptr;
        if (list1 == nullptr && list2 != nullptr) return list2;
        if (list2 == nullptr && list1 != nullptr) return list1;

        ListNode* pointer = nullptr;
        ListNode* temp1 = list1;
        ListNode* temp2 = list2;

        // Initial case for pointer setup
        if (temp1->val <= temp2->val){
            pointer = temp1;
            temp1 = temp1->next;;
        }
        else{
            pointer = temp2;
            temp2 = temp2->next;;
        }

        ListNode* result = pointer;

        while (temp1 != nullptr && temp2 != nullptr){
            if (temp1->val <= temp2->val){
                pointer->next = temp1;
                pointer = temp1;
                temp1 = temp1->next;
            }
            else{
                pointer->next = temp2;
                pointer = temp2;
                temp2 = temp2->next;
            }
        }

        if (temp1 != nullptr && temp2 == nullptr){
            pointer->next = temp1;
        }
        if (temp2 != nullptr && temp1 == nullptr){
            pointer->next = temp2;
        }

        return result;
    }
};

File: neetcode/linked_list/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 == None and list2 != None:
            return list2
        elif list2 == None and list1 != None:
            return list1
            
        fptr = list1
        sptr = list2
        newptr = None
        endptr = newptr

        while fptr != None and sptr != None:
            if fptr.val <= sptr.val:
                newNode = fptr
                fptr = fptr.next
            elif fptr.val > sptr.val:
                newNode = sptr
                sptr = sptr.next

            if not endptr:
                newptr = newNode
                endptr = newptr
            else:
                endptr.next = newNode
                endptr = endptr.next
        
        if fptr != None:
            endptr.next = fptr
        if sptr != None:
            endptr.next = sptr
        
        return newptr
